### PR TITLE
close #1352 send_all_now to all customer_notifications

### LINKED ISSUE #1352: Send_to_all_user Customer_notifications make job

### PR DESCRIPTION
**Pull request by LengTech11**
https://github.com/channainfo/commissioner/assets/119285234/2964168e-65ea-43c4-9af1-61a91626850e

### CODE REVIEW

File: app/controllers/spree/admin/customer_notifications_controller.rb [hunk 1]
```diff
@@ -29,6 +29,17 @@ def send_test
           redirect_to admin_customer_notifications_path, alert: e.message
         end
       end
+
+      def send_all
+        customer_notification = SpreeCmCommissioner::CustomerNotification.find(params[:notification_id])
+        context = { customer_notification: customer_notification, user_ids: nil }
```
Comment (channainfo) [inline app/controllers/spree/admin/customer_notifications_controller.rb:35]: let's add a new fiels call all: true. it is easy to make mistake with user_ids: nil to refer to all. we need a more explicit one ( all: true ) to send to all.

File: app/controllers/spree/admin/customer_notifications_controller.rb [hunk 2]
```diff
@@ -29,6 +29,17 @@ def send_test
...
+        begin
+          response = SpreeCmCommissioner::CustomerNotificationSender.call(context)
```
Comment (channainfo) [inline app/controllers/spree/admin/customer_notifications_controller.rb:37]: is this run in the background job?

**Review by channainfo — CHANGES_REQUESTED** [review]

File: app/controllers/spree/admin/customer_notifications_controller.rb (hunk 2 above)
Comment (LengTech11) [reply app/controllers/spree/admin/customer_notifications_controller.rb:37]: No bong. in directly deliver notification to user by call notification content by user_id and notification_id

Screenshot 2024-05-08 at 9 39 30 in the morning

File: app/controllers/spree/admin/customer_notifications_controller.rb (hunk 1 above)
Comment (LengTech11) [reply app/controllers/spree/admin/customer_notifications_controller.rb:35]: can I resolve it as like this ?

Screenshot 2024-05-08 at 9 56 59 in the morning

**Comment by LengTech11** [comment]
#### Problem
https://github.com/channainfo/commissioner/issues/1411

File: app/controllers/spree/admin/customer_notifications_controller.rb [hunk 3]
```diff
@@ -18,15 +18,21 @@ def collection_url(options = {})
...
-          redirect_to admin_customer_notifications_path, notice: I18n.t('notification.send_test_success') if response.success?
-        rescue StandardError => e
-          redirect_to admin_customer_notifications_path, alert: e.message
+        user_ids = params[:user_ids] # This can be nil or an array of user ids
+
+        if params[:send_all] == 'true'
+          SpreeCmCommissioner::CustomerNotificationSenderJob.perform_later(customer_notification.id)
+          redirect_to admin_customer_notifications_path, notice: I18n.t('notification.send_all_in_progress')
+        else
+          context = { customer_notification: customer_notification, user_ids: user_ids }
```
Comment (channainfo) [inline app/controllers/spree/admin/customer_notifications_controller.rb:29]: wheter it sends to all or not, we must send using the queue.

**Review by channainfo — CHANGES_REQUESTED** [review]

**Pull request merged**